general/mem_manager.cpp

```diff
@@ -77,286 +112,624 @@ MemoryManager::~MemoryManager()
    }
-   // Push
-   if (!(ptr_on_host && run_on_device)) { mfem_error("PtrKnown !(host && gpu)"); }
-   CuMemcpyHtoD(base.d_ptr, ptr, bytes);
-   base.host = false;
-   return base.d_ptr;
 }
 
-// Turn an alias into the right host or device address. Alloc, Push, or Pull it
-// if necessary.
-static void *PtrAlias(internal::Ledger *maps, void *ptr)
+static void PullAlias(const internal::Ledger *maps,
+                      const void *ptr, const std::size_t bytes, bool copy_data)
 {
-   const bool gpu = Device::Allows(Backend::DEVICE_MASK);
    const internal::Alias *alias = maps->aliases.at(ptr);
-   const internal::Memory *base = alias->mem;
-   const bool host = base->host;
-   const bool device = !base->host;
-   const std::size_t bytes = base->bytes;
-   if (host && !gpu) { return ptr; }
-   if (bytes==0) { mfem_error("PtrAlias bytes==0"); }
-   if (!base->d_ptr) { CuMemAlloc(&(alias->mem->d_ptr), bytes); }
-   if (!base->d_ptr) { mfem_error("PtrAlias !base->d_ptr"); }
-   void *a_ptr = static_cast<char*>(base->d_ptr) + alias->offset;
-   if (device && gpu) { return a_ptr; }
-   if (!base->h_ptr) { mfem_error("PtrAlias !base->h_ptr"); }
-   if (device && !gpu) // Pull
-   {
-      CuMemcpyDtoH(base->h_ptr, base->d_ptr, bytes);
-      alias->mem->host = true;
-      return ptr;
+   MFEM_ASSERT((char*)alias->mem->h_ptr + alias->offset == ptr,
+               "internal error");
+   // There are cases where it is OK if alias->mem->d_ptr is not allocated yet:
+   // for example, when requesting read-write access on host to memory created
+   // as device memory.
+   if (copy_data && alias->mem->d_ptr)
+   {
+      CuMemcpyDtoH(const_cast<void*>(ptr),
+                   static_cast<char*>(alias->mem->d_ptr) + alias->offset,
+                   bytes);
    }
-   // Push
-   if (!(host && gpu)) { mfem_error("PtrAlias !(host && gpu)"); }
-   CuMemcpyHtoD(base->d_ptr, base->h_ptr, bytes);
-   alias->mem->host = false;
-   return a_ptr;
 }
 
-void *MemoryManager::Ptr(void *ptr)
+void MemoryManager::RegisterCheck(void *ptr)
 {
-   if (ptr==NULL) { return NULL; };
-   if (MmDeviceIniFilter()) { return ptr; }
-   if (IsKnown(ptr)) { return PtrKnown(maps, ptr); }
-   if (IsAlias(ptr)) { return PtrAlias(maps, ptr); }
-   if (Device::Allows(Backend::DEVICE_MASK))
+   if (ptr != NULL && UsingMM())
    {
-      mfem_error("Trying to use unknown pointer on the DEVICE!");
+      if (!IsKnown(ptr))
+      {
+         mfem_error("Pointer is not registered!");
+      }
    }
-   return ptr;
 }
 
-const void *MemoryManager::Ptr(const void *ptr)
+void MemoryManager::PrintPtrs(void)
 {
-   return static_cast<const void*>(Ptr(const_cast<void*>(ptr)));
+   for (const auto& n : maps->memories)
+   {
+      const internal::Memory &mem = n.second;
+      mfem::out << std::endl
+                << "key " << n.first << ", "
+                << "host " << mem.host << ", "
+                << "h_ptr " << mem.h_ptr << ", "
+                << "d_ptr " << mem.d_ptr;
+   }
 }
```

Comment:
Needs one extra `std::endl` at the end of the prints, otherwise the next line will be combined with the last memory print line

Reply:
I'll fix it.